### PR TITLE
filters/auth: make `oauthGrant` check token cookie is allowed for request domain

### PR DESCRIPTION
**Pull request by AlexanderYastrebov**
filters/auth: make oauthGrant check token cookie is allowed for request domain

This prevents reuse of leaked cookie in another domain

### CODE REVIEW

**Comment by szuecs** [comment]
:+1:

File: filters/auth/grantcookie.go [hunk 1]
```diff
@@ -41,41 +44,48 @@ func (c *cookie) isAccessTokenExpired() bool {
...
-			request.AddCookie(c)
+		cookie, err := decodeCookie(c.Value, config)
+		if err == nil && cookie.allowedForHost(request.Host) {
+			request.Header.Del("Cookie")
+			for _, c := range cookies[:i] {
+				request.AddCookie(c)
+			}
+			for _, c := range cookies[i+1:] {
+				request.AddCookie(c)
+			}
```
Comment (szuecs) [inline filters/auth/grantcookie.go:79]: Tricky the two loops to spot that you drop element i

File: filters/auth/grantcookie.go [hunk 2]
```diff
@@ -41,41 +44,48 @@ func (c *cookie) isAccessTokenExpired() bool {
...
 		}
-		new = append(new, c)
-	}
 
-	if cookie != nil {
-		request.Header.Del("Cookie")
-		for _, c := range new {
-			request.AddCookie(c)
+		cookie, err := decodeCookie(c.Value, config)
+		if err == nil && cookie.allowedForHost(request.Host) {
```
Comment (szuecs) [inline filters/auth/grantcookie.go:72]: Here we could also continue on err != nil
Which is maybe more obvious than the current code

**Review by szuecs — COMMENTED** [review]
Feel free to merge looks good only small nits

File: filters/auth/grantcookie.go (hunk 1 above)
Comment (AlexanderYastrebov) [reply filters/auth/grantcookie.go]: Thank you for pointing this!
I've refactored it into a single loop and also noticed we had no proper test for it so I've added one

**Comment by szuecs** [comment]
:+1:

**Comment by AlexanderYastrebov** [comment]
:+1:

**Pull request merged**